src/app/api/v1/hibp/notify/route.ts

```diff
@@ -2,43 +2,84 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
+import { captureException, captureMessage } from "@sentry/node";
+
 import { NextRequest, NextResponse } from "next/server";
 import { bearerToken } from "../../../utils/auth";
 
-import { PubSub } from "@google-cloud/pubsub";
-
 const projectId = process.env.GCP_PUBSUB_PROJECT_ID;
 const topicName = process.env.GCP_PUBSUB_TOPIC_NAME;
 const subscriptionName = process.env.GCP_PUBSUB_SUBSCRIPTION_NAME;
 
+type BreachAlert = {
+  breachName: string;
+  hashPrefix: string;
+  hashSuffixes: string[];
+};
+
 /**
  * Whenever a breach is detected on the HIBP side, HIBP sends a request to this endpoint.
  * The payload is checked for validity, and immediately queued if it is valid.
  *
- * @param req
+ * @param req {NextRequest} - application/x-www-form-urlencoded POST request containing JSON.
  */
 export async function POST(req: NextRequest) {
-  if (!projectId) {
-    throw new Error("GCP_PUBSUB_PROJECT_ID env var not set");
-  }
-  if (!topicName) {
-    throw new Error("GCP_PUBSUB_TOPIC_NAME env var not set");
-  }
-
   const headerToken = bearerToken(req);
   if (headerToken !== process.env.HIBP_NOTIFY_TOKEN) {
     return NextResponse.json({ success: false }, { status: 401 });
   }
 
-  const json = await req.json();
+  const breachAlert: BreachAlert = await req.json();
 
-  if (!(json.breachName && json.hashPrefix && json.hashSuffixes)) {
+  if (
+    !(
+      breachAlert.breachName &&
+      breachAlert.hashPrefix &&
+      breachAlert.hashSuffixes
+    )
+  ) {
     console.error(
       "HIBP breach notification: requires breachName, hashPrefix, and hashSuffixes."
     );
     return NextResponse.json({ success: false }, { status: 400 });
   }
 
+  // Skip pubsub altogether for Heroku, process breach and send email immediately.
+  if (process.env.APP_ENV === "heroku") {
+    const receivedMessages = [
+      {
+        message: {
+          data: Buffer.from(JSON.stringify(breachAlert)),
+        },
+      },
+    ];
+    const { poll } = await import("../../../../../scripts/emailBreachAlerts");
+    const subClient = {
+      subscriptionPath: () => "",
+      acknowledge: (ackId: string) => console.log("acknowledged:", ackId),
+    };
+
+    // @ts-expect-error Mocking SubscriberClient
+    await poll(subClient, receivedMessages);
+
+    return NextResponse.json({ success: true }, { status: 200 });
+  }
+
+  // Otherwise, use pubub or emulator (@see README)
+  if (!projectId) {
+    const ex = "GCP_PUBSUB_PROJECT_ID env var not set, message lost";
+    captureMessage(JSON.stringify(breachAlert));
+    captureException(ex);
+    throw new Error(ex);
+  }
+  if (!topicName) {
+    const ex = "GCP_PUBSUB_TOPIC_NAME env var not set, message lost";
+    captureMessage(JSON.stringify(breachAlert));
+    captureException(ex);
+    throw new Error(ex);
+  }
+
+  const { PubSub } = await import("@google-cloud/pubsub");
   const pubsub = new PubSub({ projectId });
```

Comment:
I was discussing this with @bbangert earlier and he suggested that we cache this and the topic after lookup instead of re-creating the connection every time this function is called ... that'll be a little awkward with this new Heroku "short-circuit" approach but probably worth it.